[RLlib] Docs do-over (new API stack): Add scaling guide rst page. (#49528)
ray-project/ray: doc/source/rllib/scaling-guide.rst

Set the number of remote :py:class:`~ray.rllib.core.learner.learner.Learner` actors through:

.. testcode::

    from ray.rllib.algorithms.ppo import PPOConfig

    config = (
        PPOConfig()
        # Use 2 remote Learner actors (default is 0) for distributed data parallelism.
        # Choosing 0 creates a local Learner instance on the main Algorithm process.
        .learners(num_learners=2)
    )

Normally, you use as many :py:class:`~ray.rllib.core.learner.learner.Learner` actors as you have GPUs available for training.
Make sure, you set the number of GPUs per :py:class:`~ray.rllib.core.learner.learner.Learner` to 1:

.. testcode::

    config.learners(num_gpus_per_learner=1)

.. warning::
    For some algorithms, such as IMPALA and APPO, the performance of a single remote
    :py:class:`~ray.rllib.core.learner.learner.Learner` actor (``num_learners=1``) vs a
    single local :py:class:`~ray.rllib.core.learner.learner.Learner` instance (``num_learners=0``),
    depends on whether you have a GPU available or not.
    If exactly one GPU is available, you should run these two algorithms with ``num_learners=0, num_gpus_per_learner=1``,

>     if no GPU is available, you should set ``num_learners=1, num_gpus_per_learner=0``. For > 1 GPUs available, you should

Reviewer: ```suggestion
    if no GPU is available, you should set ``num_learners=1, num_gpus_per_learner=0``. If more than 1 GPU is available,
```

>     if no GPU is available, you should set ``num_learners=1, num_gpus_per_learner=0``. For > 1 GPUs available, you should

Reviewer: ```suggestion
    if no GPU is available, set ``num_learners=1, num_gpus_per_learner=0``. For > 1 GPUs available, you should
```

    set ``num_learners=.., num_gpus_per_learner=1``.

The number of GPUs may be fractional quantities, for example 0.5, to allocate only a fraction of a GPU per
:py:class:`~ray.rllib.env.env_runner.EnvRunner`. For example, you can pack five :py:class:`~ray.rllib.algorithms.algorithm.Algorithm`
instances onto one GPU by setting ``num_learners=1, num_gpus_per_learner=0.2``.

> See `this fractional GPU example here <https://github.com/ray-project/ray/blob/master/rllib/examples/gpus/fractional_gpus.py>`__

Reviewer: ```suggestion
See this `fractional GPU example <https://github.com/ray-project/ray/blob/master/rllib/examples/gpus/fractional_gpus.py>`__
```

for details.

.. note::
    If you specify ``num_gpus_per_learner > 0`` and your machine doesn't have the required number of GPUs
    available, the experiment may stall until the Ray autoscaler brings up enough machines to fulfill the resource request.

>     If your cluster has autoscaling turned off, this then results in a seemingly hanging experiment run.

Reviewer: ```suggestion
    If your cluster has autoscaling turned off, this setting then results in a seemingly hanging experiment run.
```

>     On the other hand, if you set ``num_gpus_per_learner=0``, RLlib builds your :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`

Reviewer: ```suggestion
    On the other hand, if you set ``num_gpus_per_learner=0``, RLlib builds the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`
```

    instances solely on CPUs, even if GPUs are available on the cluster.

Outlook: More RLlib elements that should scale
----------------------------------------------

There are other components and aspects in RLlib that should be able to scale up.

> For example, the model size is currently limited to what ever fits on a single GPU, due to

Reviewer: ```suggestion
For example, the model size is limited to whatever fits on a single GPU, due to
```
